Add statistics aggregation for the secondary source.

bbockelm opened the pull request:
This breaks out the FJR information (and corresponding HTCondor statistics) for the secondary source versus all sources.

The intent is to provide insight for amount of input data for PileUp-vs-Primary in various DIGI-RECO workflows.

@osschar

jbsauvan committed:
Add approximate Zoltan/Split geometry

jbsauvan committed:
8 inch trigger cell mapping fixes

jbsauvan committed:
Add thickness equalization in frontend

jbsauvan committed:
Fix configs

jbsauvan committed:
Remove thickness-dependent corrections from the backend

jbsauvan committed:
Update wafer ordering

mastrolorenzo committed:
fix cell-thickness correction in HGCalTriggerCellCalibration.cc(.h)

jbsauvan committed:
Add V8 neighbors

jbsauvan committed:
Update configs

jbsauvan committed:
Add trigger geometry customization

amarini committed:
genjets HGC L1 Trigger Nutples

amarini committed:
jb comments.

jbsauvan committed:
Add TC links in 2D cluster ntuple

jbsauvan committed:
Add 2D cluster links to 3D clusters

mastrolorenzo committed:
added ntuplizer for gen-tau + L1Clustering_taus_cfg.py to generate single taus with Pythia8

mastrolorenzo committed:
update geometry modules in L1Clustering_cfg.py: from 2023D4 to 2023D17

mastrolorenzo committed:
update in HGCalTriggerNtupleGenTau.cc

bbockelm committed:
Make compression settings more aggressive.

Summary:
- Keep LZMA-4 as a default for data tiers that are not transient.
- Change default auto-flush to 20MB (currently 10MB).
- Change auto-flush for *AOD* to 30MB (currently 15MB).

Details:
Did a bit of investigating inside ROOT compression settings (which maybe hasn't been done since Run I?)

I took a file (/store/data/Run2017A/SingleMuon/AOD/PromptReco-v3/000/296/888/00000/002DA5DA-6455-E711-9F13-02163E014259.root) from what should be the most prevalent data tier (AOD) and recompressed it under a variety of settings (ZLIB-7, ZLIB-9, LZMA-4, LZMA-9).  A few findings.

Relative speed*:
- ZLIB-7: 15MB/s, 0.021s/evt
- ZLIB-9: 2.7MB/s, 0.11s/evt
- LZMA-4: 1.9MB/s, 0.15s/evt
- LZMA-9: 1.1MB/s, 0.26s/evt

Nothing new here.  LZMA-9 is about an order-magnitude slower than ZLIB-7; this is consistent with prior measurements.

The default compression settings for this file is LZMA-4 with 15MB auto-flush buffer, resulting in a 2,816MB file.

Relative file sizes (10MB auto-flush buffer):
- ZLIB-7: 3,150MB
- ZLIB-9: 3,129MB
- LZMA-4: 2,925MB, (+3.9% file size increase, 26MB RSS buffer savings)
- LZMA-9: 2,891MB, (+2.7%, 19MB RSS buffer savings)

Relative file sizes (20MB auto-flush buffer):
- ZLIB-7: 2,996MB
- ZLIB-9: 2,974MB
- LZMA-4: 2,739MB (-2.7% smaller than original, additional 29MB RSS in buffer usage)
- LZMA-9: 2,701MB (-4.1% smaller, additional 23MB RSS)

Using a 30MB auto-flush buffer:
- ZLIB-7: 2,941
- ZLIB-9: 2,918
- LZMA-4: 2,666 (-5.3% smaller, +91MB RSS)
- LZMA-9: 2,626 (-6.5% smaller, +94MB RSS)

So, for about 90MB per output file, we can get 5% additional file size savings.

Other thoughts:
- With multithreading, we have additional room in the memory budget.
- Until ROOT 6.12 is available (which multithreads compression using TBB), we ought to be careful about significantly increasing CPU usage per-event for output, is it can cause significant stalls.

jbsauvan committed:
Add neighbors in layer-based geometry

mastrolorenzo committed:
1st simplification of HGCalTriggerNtupleGenTau + cleaning

jbsauvan committed:
Fix neighbor reading

jbsauvan committed:
Fix indentations

jbsauvan committed:
Add neighbor file for Zoltan/Split geometry

jbsauvan committed:
Add Zoltan/Split mapping for the V8 geometry

jbsauvan committed:
Add neighbor distance in geometry tree

jbsauvan committed:
Update geometry configs

mastrolorenzo committed:
minor fixes

mastrolorenzo committed:
update with statusFlags to select tau decays

jbsauvan committed:
Add missing energy equalization in BestChoice codec

mastrolorenzo committed:
updates in HGCalTriggerNtupleGenTau.cc + cleaning

mastrolorenzo committed:
removed L1Clustering_taus_cfg.py

mastrolorenzo committed:
removed ntuple_gentau from the standard sequence in hgcalTriggerNtuples_cfi.py

jbsauvan committed:
Add BH in trigger geometry

jbsauvan committed:
Update panel numbering scheme

jbsauvan committed:
Add BH in front-end codecs

jbsauvan committed:
Fix bestchoice codec

jbsauvan committed:
Disconnect trigger modules with id 0

jbsauvan committed:
Add neighbors in the BH

jbsauvan committed:
fix BH threshold

jbsauvan committed:
Add BH in clustering

jbsauvan committed:
Add BH trigger cell calibration

jbsauvan committed:
Add separate cluster threshold for BH

jbsauvan committed:
Remove obsolete geometry data

jbsauvan committed:
Remove unecessary dependencies in BuildFile

jbsauvan committed:
Add selection of endcap HCAL cells

jbsauvan committed:
Add BH simhit matching

jbsauvan committed:
Disconnect BH in the old geometry

jbsauvan committed:
Fix fe testers

jbsauvan committed:
Revert input file in test config

jbsauvan committed:
Update configs for 930

jbsauvan committed:
Clean geometry configs

jbsauvan committed:
Remove data files

jbsauvan committed:
Merged hgc-tpg-devel-CMSSW_9_3_0_pre2 from repository PFCal-dev

civanch committed:
substitute auto_ptr and remove redundant code

civanch committed:
cleanup initialisation

civanch committed:
more cleanup

jbsauvan committed:
Fix indents

jbsauvan committed:
Remove hardcoded values

jbsauvan committed:
Add const function identifier

jbsauvan committed:
Customize MaxCellsInModule in geometry custom functions

jbsauvan committed:
Use Hcal id relabeller

jbsauvan committed:
Use exceptions instead of log errors

jbsauvan committed:
Move disconnected modules handling in geometry

jbsauvan committed:
Add geometries README

civanch committed:
cosmetics

civanch committed:
fixed typo

civanch committed:
fixed runMatrix error

civanch committed:
cleanup

Commit:
Added check that trackAssociator exists before running - an exception is called if it does not exist. Updated plots - removed module level plots and replaced those with 1D plots per layer/disk. Maps per module per ladder/blade now eixst where appropriate for module level information without significantly increasing the budget akin to the old 1D module level plots.

Commit:
update ToF binning following chat with experts.

ianna committed:
DetectorDescrition and Geometry Clang readability-container-size-empty

mandrenguyen committed:
Fix for mixing workflows that use playback

ianna committed:
mergea Please enter the commit message for your changes. Lines starting

ianna committed:
remove conflict

vanbesien committed:
Fixed irrational range and spaces in histo labels

DryRun committed:
Use all samples if samples < ncapid

Commit:
for neutral exotic WIMPs - send their daughters to Geant4 instead of the exotic itself

rmanzoni committed:
apply selections but isolation at the tau building step, sort taus by pt, fix isolation condition logic

rmanzoni committed:
make produce abide to the promise of being const

civanch committed:
fixed initialisation

smuzaffar committed:
added .clang-tidy

kpedro88 committed:
update HCAL trigger mode for Run3

bbockelm committed:
Add statistics aggregation for the secondary source.

This breaks out the FJR information (and corresponding HTCondor
statistics) for the secondary source versus all sources.

The intent is to provide insight for amount of input data for
PileUp-vs-Primary in various DIGI-RECO workflows.

cmsbuild commented:
@bbockelm, CMSSW_9_3_X branch is closed for direct updates. cms-bot is going to move this PR to master branch.
In future, please use cmssw master branch to submit your changes.

cmsbuild commented:
A new Pull Request was created by  @bbockelm (Brian Bockelman) for master.

It involves the following packages:

FWCore/Services
IOPool/Input
Utilities/StorageFactory


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit this is something you requested to watch as well.
@davidlange6 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

bbockelm committed:
Merge branch 'master' into secondary_source_stats

davidlange6 committed:
Merge pull request #20099 from smuzaffar/clang-tidy-file

added .clang-tidy

cmsbuild committed:
Merge pull request #20104 from kpedro88/Run3-WF2

update HCAL trigger mode for Run3

cmsbuild committed:
Merge pull request #20096 from civanch/fixed_stacking_action

Fixed initialisation of Geant4 stacking action

cmsbuild committed:
Merge pull request #20078 from DryRun/CMSSW_9_3_0_pre3_fix_hcalcalib

HCALDQM: Fix missing HF in hcalcalib

cmsbuild committed:
Merge pull request #20056 from vanbesien/Fixing_offline_HV_plot

Fixed range and unneeded spaces in histo labels

cmsbuild committed:
Merge pull request #20020 from ianna/dd-geom-clang-readability-container-size-empty

DetectorDescrition and Geometry Clang readability-container-size-empty

davidlange6 committed:
Merge pull request #19589 from bbockelm/change_flush_settings

Make compression settings more aggressive.

cmsbuild committed:
Merge pull request #19936 from PFCal-dev/hgc-tpg-integration-170727

HGCAL trigger updates (geometry, BH)

cmsbuild committed:
Merge pull request #19992 from davidcarbonis/phase1pixelValidation_930_fixes_v2

PixelPhase Validation bin reduction and protection against memory leak.

davidlange6 committed:
Revert "Ss threadsafe ecal digi to raw"

fabferro committed:
adding patch to prevent corrupted row/col from crashing

fabferro committed:
removing extra ;

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pr-code-checks/PR-20107/33

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/jenkins-artifacts/pr-code-checks/PR-20107/33/git-diff.patch

In future, you can run `scram build code-checks` to apply code checks

smuzaffar commented:
@davidlange6 , does message like https://github.com/cms-sw/cmssw/pull/20107#issuecomment-321603667 looks reasonable for code checks?

bbockelm commented:
@smuzaffar - message looks reasonable to me!

davidlange6 commented:
"in the future" means you can't do it yet?

> On Aug 10, 2017, at 6:30 PM, Malik Shahzad Muzaffar <notifications@github.com> wrote:
> 
> @davidlange6 , does message like #20107 (comment) looks reasonable for code checks?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

smuzaffar commented:
humm... I meant for future/new PRs one can run code-checks (which is already available in 93X IBs)

fabferro committed:
turning magic numbers into constants

davidlange6 commented:
so its even available for current PRs? (or does it depend on releases with a .clang-tidy?)

> On Aug 10, 2017, at 6:45 PM, Malik Shahzad Muzaffar <notifications@github.com> wrote:
> 
> humm... I meant for future/new PRs one can run code-checks (which is already available in 93X IBs)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Dr15Jones commented:
I see I am wrong about the lack of a guard.

Dr15Jones commented on FWCore/Services/plugins/CondorStatusUpdater.cc:
```diff
@@ -338,6 +339,24 @@ CondorStatusService::updateImpl(time_t sinceLastUpdate)
+    {
+      // Given there is some cost to reporting empty numbers, only push forward the secondary source info
+      // if one file was actually opened via such a source.
+      auto label = StorageAccount::setLabel("secondarysource");
```
Better to create constants somewhere with this same since you are requiring multiple places to use the same labels. Also true for the 'tokenForStorageClassName' call.

Dr15Jones commented on Utilities/StorageFactory/interface/StorageAccount.h:
```diff
@@ -116,21 +153,61 @@ class StorageAccount {
+  private:
+    OpenLabelToken(const std::string &label) {
+      m_label = label;
+    }
```
This should take the previous value of the thread local and reset it in the destructor. That way if TBB tasks call other TBB tasks the code will be reset.

Dr15Jones commented on Utilities/StorageFactory/interface/StorageAccount.h:
```diff
@@ -116,21 +153,61 @@ class StorageAccount {
+
+    static const std::string &getLabel() {return m_label;}
+
+    static thread_local std::string m_label;
```
`s_label` since it is static.

bbockelm commented:
Yeah, I'm not too thrilled about having a thread-local here either.  However, this is the only mechanism I could see to pass context from the CMSSW layer, through ROOT, and back to CMSSW.

If you have a better mechanism, I'd be happy to hear!  Otherwise, I plan on starting to work on the other review comments.

Dr15Jones commented on Utilities/StorageFactory/src/StorageAccount.cc:
```diff
@@ -50,23 +56,50 @@ static std::string d2str(double d) {
+  }
+  for (auto const &protocol : protocols) {
+    StorageAccount::setLabel("");
+    auto token = tokenForStorageClassName(protocol);
```
I think `tokenForStorageClassName` should be changed to take both the protocol and the label. Then a new function should be added `tokenForStorageClassNameUsingContext` which takes only the protocol and gets the label from the thread local. So all existing calls to `tokenForStorageClassName` would need to be changed to the new name.

Dr15Jones commented on Utilities/StorageFactory/interface/StorageAccount.h:
```diff
@@ -116,21 +153,61 @@ class StorageAccount {
+    static thread_local std::string m_label;
+  };
+
+  static OpenLabelToken setLabel(const std::string &label) {return OpenLabelToken(label);}
```
I think this should be changed to `setContextLabel` and `getContextLabel` to better explain the purpose.

davidlange6 committed:
Merge pull request #20119 from cms-sw/revert-18322-ss_threadsafe_EcalDigiToRaw

Revert "Ss threadsafe ecal digi to raw"

Dr15Jones commented on FWCore/Services/plugins/CondorStatusUpdater.cc:
```diff
@@ -338,6 +339,24 @@ CondorStatusService::updateImpl(time_t sinceLastUpdate)
+      // if one file was actually opened via such a source.
+      auto label = StorageAccount::setLabel("secondarysource");
+      const auto token = StorageAccount::tokenForStorageClassName("tstoragefile");
+      auto &operations = stats[token.value()];
```
Changing `StorageAccount::summary` to return a non-const object just so you can call `operator[]` does not seem reasonable to me. You could just call `find` on the const object.

smuzaffar commented:
@davidlange6 , it is only available for latest 93X IBs only. code-checks build rules uses .clang-tidy file if available otherwise it fallback to use the checks we used to create .clang-tidy.

For now this check is only run for PRs made against master branch. If we want to run it for older IBs (92X, 91X ) then we have to updated those release cycles to use latest build rules.

bbockelm committed:
Fix C++11 items from clang-tidy.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pr-code-checks/PR-20107/47

cmsbuild committed:
Merge pull request #20120 from CTPPS/ctpps_pixel_unpacker_patch1_93x

CTPPS Pixel unpacker patch

cmsbuild committed:
Merge pull request #20088 from rmanzoni/master_updatesHLTtau3mu

updates to tau3mu HLT producer

cmsbuild committed:
Merge pull request #20083 from zhangzc11/93X_dev_longLived_onlyDaughterToGeant4

for neutral exotic WIMPs - send their daughters to Geant4 instead of …

cmsbuild committed:
Merge pull request #20024 from mandrenguyen/playbackBugFix93X

Fix mixing workflows that use playback

cmsbuild committed:
Merge pull request #19939 from civanch/cleanup_g4_sensitive_detectors

Cleanup Geant4 sensitive detectors

bbockelm committed:
Merged refs/pull/20107/head from repository cms-sw

bbockelm commented:
Noted a potential problem here while implementing @Dr15Jones's suggestions -- the `TStorageFactoryFile` does not cache the token, meaning it will look up a new one each read (potentially changing the context).  Will include a fix for this with the other changes.

bbockelm committed:
Switch labels to an enumerated, more explicit context.

This turns the labels from std::strings to an enum class.  Changes
the API to allow the context to be seen more explicitly.

bbockelm commented:
Arrrgh.  Merging confused GitHub UI.  Will also fix that on Monday.

cmsbuild commented:
Pull request #20107 was updated. @ianna, @kpedro88, @Martin-Grunewald, @rekovic, @vanbesien, @perrotta, @civanch, @silviodonato, @cmsbuild, @fwyzard, @davidlange6, @smuzaffar, @Dr15Jones, @mdhildreth, @slava77, @mommsen, @mulhearn, @vazzolini, @kmaeshima, @emeschi, @dmitrijus, @franzoni can you please check and sign again.

bbockelm commented:
Actually, since cmsbuild tagged a bunch of other groups, the PR's state will be hopelessly confused.  Will close and re-open a clean branch.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pr-code-checks/PR-20107/74

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/jenkins-artifacts/pr-code-checks/PR-20107/74/git-diff.patch

In future, you can run `scram build code-checks` to apply code checks
